react-app/src/components/TaskGrid.js

```diff
@@ -1,31 +1,56 @@
-import React, { useEffect, useRef, useState } from 'react'
-import { createTask } from '../domain/Task'
-import { createTaskHolder, DESC } from '../domain/TaskHolder'
-import TaskCard from './TaskCard';
+import React, { useEffect, useState } from 'react'
+import { v4 as uuidv4 } from 'uuid';
+import { complete, createTask } from '../domain/Task'
+import FlipTaskCard from './FlipTaskCard'
 
-function TaskGrid({}) {
-    const taskHolder = useRef(null)
+export const REMOVE_TASK_DELAY = 1000
+
+function TaskGrid() {
     const [tasks, setTasks] = useState([])
 
     useEffect(() => {
-        taskHolder.current = createTaskHolder()
-            .add(createTask({ title: 'Find dad\'s wallet', rewardPoints: 125 }))
-            .add(createTask({ title: 'Put away old toys to white boxes on the balconey', rewardPoints: 100 }))
-            .add(createTask({ title: 'Wash the dishes', rewardPoints: 75 }))
-            .add(createTask({ title: 'Make your bed in the morning', rewardPoints: 20 }))
-            .add(createTask({ title: 'Fix the pillowcase', rewardPoints: 50 }))
-            .add(createTask({ title: 'Water the plants', rewardPoints: 80 }))
-            .sortBy(DESC)
-
-        const { tasks } = taskHolder.current
-        setTasks(tasks)
-    }, [])
+        setTasks([
+            createTask({ id: uuidv4(), title: 'Put away old toys to white boxes on the balconey', rewardPoints: 100 }),
+            createTask({ id: uuidv4(), title: 'Find dad\'s wallet', rewardPoints: 125 }),
+            createTask({ id: uuidv4(), title: 'Wash the dishes', rewardPoints: 75 }),
+            createTask({ id: uuidv4(), title: 'Make your bed in the morning', rewardPoints: 20 }),
+            createTask({ id: uuidv4(), title: 'Fix the pillowcase', rewardPoints: 50 }),
+            createTask({ id: uuidv4(), title: 'Water the plants', rewardPoints: 80 }),
+        ]);
+
+    }, []);
+
+    function completeTask(completedTask) {
+        const newTasks = tasks.map(function completeSingleTask(t) {
+
+            if(t.id === completedTask.id) {
+                t = complete(t)
+            }
+
+            return t
+        })
```

Comment:
I would suggest you splitting the completeTask operation into 2 different steps and :

```js
// 1. 
// 2. providing a callback to remove task after some time
// 3. setting the state

function completeTask(tasks, completingTaskId) {
  // Make sure existent tasks are not mutated
  const tasksWithCompletedTask = tasks.map(function completeSingleTask(t) {
    if(t.id === completingTaskId) {
      t = complete(t);
    }

    return t;
  });

  setTasks(tasksWithCompletedTask);
};

function removeCompletedTask(tasks, completedTaskId) {
  const filteredTasks = tasks.filter(function getNonCompletedTask(task) {
    return task.id !== completedTaskId;
  });

  setTasks(filteredTasks);
};

completeTask(
  tasks, 
  task.id,
);

setTimeout(() => removeCompeletedTask(
  tasks, 
  task.id,
), DELAY);  
```

This way it's a bit easier to understand the process of task completion (i.e. we first complete the task and after that schedule the removal).

Alternatively, you can put delayed callback call in complete task, but imo it is unnecessary here:

```js
function completeTask(tasks, completingTaskId, { onCompleteDelayed }) {
  // Make sure existent tasks are not mutated
  const tasksWithCompletedTask = tasks.map(function completeSingleTask(t) {
    if(t.id === completingTaskId) {
      t = complete(t);
    }

    return t;
  });

  setTimeout(() => onCompleteDelayed(
    tasksWithCompletedTask, completingTaskId
  ), DELAY);
  setTasks(tasksWithCompletedTask);
};

// ...

completeTask(
  tasks, 
  task.id,
  { onCompleteDelayed: () => 
    removeCompeletedTask(
      tasks, 
      task.id,
    ),
  },
);
```

But in both solutions there might be incosistency as the list of tasks i.e. `tasksWithCompletedTask` may be outdated (i.e. other tasks were completed by the time the timer starts running), so I suggest you to re-check that there's no bug.

Reply:
I will try implementing it when using the firebase sdk.